src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/SpecialFileProviders/AbstractSpecialFileProvider.cs

```diff
@@ -0,0 +1,261 @@
+        /// </summary>
+        protected virtual bool CreatedByDefaultUnderAppDesignerFolder => true;
+
+        /// <summary>
+        /// Gets the name of a special file.
+        /// </summary>
+        protected abstract string GetFileNameOfSpecialFile(SpecialFiles fileId);
+
+        /// <summary>
+        /// Gets the name of a template that can be used to create a new special file.
+        /// </summary>
+        protected abstract string GetTemplateForSpecialFile(SpecialFiles fileId);
+
+        /// <summary>
+        /// We follow this algorithm for looking up files:
+        ///
+        /// if (not asked to create)
+        ///      Look in AppDesigner folder
+        ///      Look in root folder
+        ///
+        /// if (asked to create)
+        ///      Look in AppDesigner folder
+        ///      Look in root folder
+        ///      Force-create in app-designer folder unless it's the app.config file, which
+        ///          the users expect in the root folder.
+        /// </summary>
+        public async Task<string> GetFileAsync(SpecialFiles fileId, SpecialFileFlags flags, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            string specialFileName = GetFileNameOfSpecialFile(fileId);
+
+            // Search for the file in the app designer and root folders.
+            IProjectTree specialFileNode = FindFile(specialFileName);
+            if (specialFileNode != null)
+            {
+                if (await IsNodeInSyncWithDiskAsync(specialFileNode, forceSync: flags.HasFlag(SpecialFileFlags.CreateIfNotExist), cancellationToken: cancellationToken))
+                {
+                    return specialFileNode.FilePath;
+                }
+            }
+
+            // File doesn't exist. Create it if we've been asked to.
+            if (flags.HasFlag(SpecialFileFlags.CreateIfNotExist))
+            {
+                string createdFilePath = await CreateFileAsync(fileId, specialFileName);
+                if (createdFilePath != null)
+                {
+                    return createdFilePath;
+                }
+            }
+
+            // We haven't found the file but return the default file path as that's the contract.
+            IProjectTree rootNode = _projectTreeService.CurrentTree.Tree;
+            string rootFilePath = _projectTreeService.CurrentTree.TreeProvider.GetPath(rootNode);
+            string fullPath = Path.Combine(rootFilePath, specialFileName);
+            return fullPath;
+        }
+
+        /// <summary>
+        /// Find a file with the given file name. The algorithm used is :
+        ///       Look under the appdesigner folder for files that normally live there.
+        ///       Look under the project root for all files.
+        /// </summary>
+        private IProjectTree FindFile(string specialFileName)
+        {
+            IProjectTree rootNode = _projectTreeService.CurrentTree.Tree;
+
+            if (specialFileName == null)
```

Comment:
Why would we be passed a empty special filename? Wouldn't that be a bug in the implementor?